File: tetris.js
#!/usr/bin/env nodejs

var PLAYFIELD_W = 10,
    PLAYFIELD_H = 20,
    PLAYFIELD_X = 30,
    PLAYFIELD_Y = 1,
    BORDER_COLOR = "yellow",

    HELP_X = 58,
    HELP_Y = 1,
    HELP_COLOR = "cyan",

    SCORE_X = 1,
    SCORE_Y = 2,
    SCORE_COLOR = "green",

    NEXT_X = 14,
    NEXT_Y = 11,

    GAMEOVER_X = 1,
    GAMEOVER_Y = PLAYFIELD_H + 3,

    INITIAL_MOVE_DOWN_DELAY = 1000,
    DELAY_FACTOR = 0.8,
    LEVEL_UP = 20,

    NEXT_EMPTY_CELL = "  ",
    PLAYFIELD_EMPTY_CELL = " .",
    FILLED_CELL = "[]";

function TetrisScreen() {
    this.s = "";
    this.use_color = true;
}

TetrisScreen.prototype.toggle_color = function() {
    this.use_color ^= true;
}

TetrisScreen.prototype.print = function(s) {
    this.s += s;
}

TetrisScreen.prototype.flush = function() {
    process.stdout.write(this.s);
    this.s = "";
}

TetrisScreen.prototype.reset_colors = function() {
    this.s += "\u001B[0m"
}

TetrisScreen.prototype.set_bold = function() {
    this.s += "\u001B[1m"
}

TetrisScreen.prototype.clear_screen = function() {
    this.s += "\u001B[2J"
}

TetrisScreen.prototype.show_cursor = function() {
    this.s += "\u001B[?25h"
}

TetrisScreen.prototype.hide_cursor = function () {
    this.s += "\u001B[?25l"
}

TetrisScreen.color = {
    'red'     : 1,
    'green'   : 2,
    'yellow'  : 3,
    'blue'    : 4,
    'fuchsia' : 5,
    'cyan'    : 6,
    'white'   : 7
};

TetrisScreen.prototype.xyprint = function(x, y, s) {
    this.s += ("\u001B[" + y + ";" + x + "H" + s);
}

TetrisScreen.prototype.set_fg = function (c) {
    if (this.use_color) {
        this.s += ("\u001B[3" + TetrisScreen.color[c] + "m");
    }
}

TetrisScreen.prototype.set_bg = function (c) {
    if (this.use_color) {
        this.s += ("\u001B[4" + TetrisScreen.color[c] + "m");
    }
}

TetrisScreen.prototype.get_random_color = function() {
    a = Object.keys(TetrisScreen.color);
    return a[Math.floor(Math.random() * a.length)];
}

function TetrisPlayField(screen) {
    this.screen = screen;
    this.cells = [];
    for (var i = 0; i < PLAYFIELD_H; i++) {
        this.cells[i] = [];
    }
}

TetrisPlayField.prototype.draw_border = function() {
    var blocks = ['==', '\\/'];
    this.screen.set_bold()
    this.screen.set_fg(BORDER_COLOR)
    for (var i = 0; i < PLAYFIELD_H; i++) {
        var y = i + PLAYFIELD_Y;
        this.screen.xyprint(PLAYFIELD_X - 2, y, "<|");
        this.screen.xyprint(PLAYFIELD_X + PLAYFIELD_W * 2, y, "|>");
    }
    for (var i = 0; i < blocks.length; i++) {
        this.screen.xyprint(PLAYFIELD_X, PLAYFIELD_Y + PLAYFIELD_H + i, Array(PLAYFIELD_W + 1).join(blocks[i]));
    }
    this.screen.reset_colors()
}

TetrisPlayField.prototype.show = function() {
    for (var y = 0; y < PLAYFIELD_H; y++) {
        this.screen.xyprint(PLAYFIELD_X, PLAYFIELD_Y + y, "");
        for (var x = 0; x < PLAYFIELD_W; x++) {
            var cell = this.cells[y][x];
            if (cell) {
                this.screen.set_fg(cell);
                this.screen.set_bg(cell);
                this.screen.print(FILLED_CELL);
                this.screen.reset_colors();
            } else {
                this.screen.print(PLAYFIELD_EMPTY_CELL);
            }
        }
    }
}

TetrisPlayField.prototype.position_ok = function(piece, position) {
    return piece.get_cells(position).every(function(cell) {
        var x = cell.x;
        var y = cell.y;
        return (x >= 0 && x < PLAYFIELD_W && y >= 0 && y < PLAYFIELD_H && this.cells[y][x] == undefined);
    }, this);
}

TetrisPlayField.prototype.flatten_piece = function(piece) {
    piece.get_cells().forEach(function(cell) {
        this.cells[cell.y][cell.x] = piece.color;
    }, this);
}

TetrisPlayField.prototype.process_complete_lines = function() {
    var cells = this.cells.filter(function(row) {
        for (var i = 0; i < PLAYFIELD_W; i++) {
            if (row[i] == undefined) {
                return true;
            }
        }
        return false;
    }, this);
    var complete_lines = PLAYFIELD_H - cells.length;
    for (var i = 0; i < complete_lines; i++) {
        cells.unshift([]);
    }
    this.cells = cells;
    return complete_lines;
}

function TetrisScreenItem(screen) {
    this.visible = true;
    this.screen = screen;
}

TetrisScreenItem.prototype.show = function() {
    if (this.visible) {
        this.draw(true);
    }
}

TetrisScreenItem.prototype.hide = function() {
    if (this.visible) {
        this.draw(false);
    }
}

TetrisScreenItem.prototype.toggle = function() {
    this.visible = ! this.visible;
    this.draw(this.visible);
}

TetrisScreenItem.prototype.set_visible = function(value) {
    this.visible = value;
}

TetrisHelp.prototype = new TetrisScreenItem();
TetrisHelp.prototype.constructor = TetrisHelp;

function TetrisHelp(screen) {
    this.screen = screen;
    this.color = HELP_COLOR;
    this.text = [
        "  Use cursor keys",
        "       or",
        "    s: rotate",
        "a: left,  d: right",
        "    space: drop",
        "      q: quit",
        "  c: toggle color",
        "n: toggle show next",
        "h: toggle this help"
    ]   
}

TetrisHelp.prototype.draw = function(visible) {
    this.screen.set_bold();
    this.screen.set_fg(this.color)
    this.text.forEach(function(s, i) {
        this.screen.xyprint(HELP_X, HELP_Y + i, visible ? s : Array(s.length + 1).join(' '))
    }, this);
    this.screen.reset_colors()
}

TetrisScore.prototype = new TetrisScreenItem();
TetrisScore.prototype.constructor = TetrisScore;

function TetrisScore(screen, tetris_input_processor) {
    this.tetris_input_processor = tetris_input_processor;
    this.screen = screen;
    this.score = 0;
    this.level = 1;
    this.lines_completed = 0;
}

TetrisScore.prototype.update = function(complete_lines) {
    this.lines_completed += complete_lines;
    this.score += (complete_lines * complete_lines);
    if (this.score > LEVEL_UP * this.level) {
        this.level += 1;
        this.tetris_input_processor.tick_time *= DELAY_FACTOR;
    }
    this.show();
}

TetrisScore.prototype.show = function() {
    this.screen.set_bold();
    this.screen.set_fg(SCORE_COLOR);
    this.screen.xyprint(SCORE_X, SCORE_Y,     "Lines completed: " + this.lines_completed);
    this.screen.xyprint(SCORE_X, SCORE_Y + 1, "Level:           " + this.level);
    this.screen.xyprint(SCORE_X, SCORE_Y + 2, "Score:           " + this.score);
    this.screen.reset_colors();
}

TetrisPiece.prototype = new TetrisScreenItem();
TetrisPiece.prototype.constructor = TetrisPiece;

function TetrisPiece(screen) {
    // 0123
    // 4567
    // 89ab
    // cdef
    this.piece_data = [
        ["1256"], // square
        ["159d", "4567"], // line
        ["4512", "0459"], // s
        ["0156", "1548"], // z
        ["159a", "8456", "0159", "2654"], // l
        ["1598", "0456", "2159", "a654"], // inverted l
        ["1456", "1596", "4569", "4159"]  // t
    ]

    this.screen = screen;
    this.color = this.screen.get_random_color();
    this.piece_index = Math.floor(Math.random() * this.piece_data.length);
    this.symmetry = this.piece_data[this.piece_index].length;
    this.position = {
        x: 0,
        y: 0,
        z: Math.floor(Math.random() * this.symmetry)
    }
    this.empty_cell = NEXT_EMPTY_CELL;
}

TetrisPiece.prototype.get_cells = function(new_position) {
    var result = [];
    var p = new_position || this.position;
    var data = this.piece_data[this.piece_index][p.z];
    data.split('').forEach(function(c, i) {
        var n = parseInt(c, 16);
        result[i] = {x: p.x + (n & 3), y: p.y + ((n >> 2) & 3)};
    }, this);
    return result;
}

TetrisPiece.prototype.draw = function(visible) {
    if (visible) {
        this.screen.set_fg(this.color);
        this.screen.set_bg(this.color);
    }
    this.get_cells().forEach(function(cell) {
        this.screen.xyprint(this.origin.x + cell.x * 2, this.origin.y + cell.y, visible ? FILLED_CELL : this.empty_cell);
    }, this);
    this.screen.reset_colors();
}

TetrisPiece.prototype.new_position = function(dx, dy, dz) {
    return {
        x: this.position.x + dx,
        y: this.position.y + dy,
        z: (this.position.z + dz) % this.symmetry
    }
}

function TetrisController(tetris_input_processor) {
    this.tetris_input_processor = tetris_input_processor;
    this.screen = new TetrisScreen();
    this.next_piece_visible = true;
    this.playfield = new TetrisPlayField(this.screen);
    this.help = new TetrisHelp(this.screen);
    this.score = new TetrisScore(this.screen, this.tetris_input_processor);
    this.get_next_piece();
    this.get_current_piece();
    this.redraw_screen();
    this.screen.flush();
}

TetrisController.prototype.get_next_piece = function() {
    this.next_piece = new TetrisPiece(this.screen);
    this.next_piece.origin = {x: NEXT_X, y: NEXT_Y};
    this.next_piece.set_visible(this.next_piece_visible);
    this.next_piece.show();
}

TetrisController.prototype.get_current_piece = function() {
    this.next_piece.hide();
    this.current_piece = this.next_piece;
    this.current_piece.position = {x: (PLAYFIELD_W - 4) / 2, y: 0, z: this.current_piece.position.z};
    if (! this.playfield.position_ok(this.current_piece)) {
        this.quit();
    }
    this.current_piece.set_visible(true);
    this.current_piece.empty_cell = PLAYFIELD_EMPTY_CELL;
    this.current_piece.origin = {x: PLAYFIELD_X, y: PLAYFIELD_Y};
    this.current_piece.show();
    this.get_next_piece();
}

TetrisController.prototype.redraw_screen = function() {
    this.screen.clear_screen();
    this.screen.hide_cursor();
    this.playfield.draw_border();
    [this.playfield, this.help, this.score, this.next_piece, this.current_piece].forEach(function(o) {
        o.show();
    });
}

TetrisController.prototype.process_key = function(key) {
    var commands = {
        '\u0003': "quit",
        'q': "quit",
        '\u001b[C': "cmd_right",
        'd': "cmd_right",
        '\u001b[D': "cmd_left",
        'a': "cmd_left",
        '\u001b[A': "cmd_rotate",
        's': "cmd_rotate",
        ' ': "cmd_drop",
        'h': "toggle_help",
        'n': "toggle_next",
        'c': "toggle_color"
    };
    if (commands[key]) {
        this[commands[key]]();
        this.screen.flush();
    }
}

TetrisController.prototype.quit = function() {
    this.screen.xyprint(GAMEOVER_X, GAMEOVER_Y, "Game over!");
    this.screen.xyprint(GAMEOVER_X, GAMEOVER_Y + 1, "");
    this.screen.show_cursor();
    this.screen.flush();
    clearTimeout(this.tetris_input_processor.tick_timeout);
    process.stdin.pause();
    process.exit();
}

TetrisController.prototype.process_fallen_piece = function() {
    this.playfield.flatten_piece(this.current_piece);
    var complete_lines = this.playfield.process_complete_lines();
    if (complete_lines > 0) {
        this.score.update(complete_lines);
        this.playfield.show();
    }
}

TetrisController.prototype.move = function(dx, dy, dz) {
    var new_position = this.current_piece.new_position(dx, dy, dz);
    if (this.playfield.position_ok(this.current_piece, new_position)) {
        this.current_piece.hide();
        this.current_piece.position = new_position;
        this.current_piece.show();
        return true;
    }
    return (dy == 0);
}

TetrisController.prototype.cmd_down = function() {
    var landed = this.move(0, 1, 0);
    this.screen.flush();
    if (landed) {
        return true;
    }
    this.process_fallen_piece();
    this.get_current_piece();
    return false;
}

TetrisController.prototype.cmd_right = function() {
    this.move(1, 0, 0);
}

TetrisController.prototype.cmd_left = function() {
    this.move(-1, 0, 0);
}

TetrisController.prototype.cmd_rotate = function() {
    this.move(0, 0, 1);
}

TetrisController.prototype.cmd_drop = function() {
    while(this.cmd_down()) {
    }
}

TetrisController.prototype.toggle_help = function() {
    this.help.toggle();
}

TetrisController.prototype.toggle_next = function() {
    this.next_piece.toggle();
}

TetrisController.prototype.toggle_color = function() {
    this.screen.toggle_color();
    this.redraw_screen();
}

function TetrisInputProcessor() {
    this.tick_time = INITIAL_MOVE_DOWN_DELAY;
    this.tick_timeout = null;
}

TetrisInputProcessor.prototype.run = function() {
    var tetris_controller = new TetrisController(this);

    var stdin = process.stdin;
    stdin.setRawMode(true);
    stdin.resume();
    stdin.setEncoding('utf8');
    var self = this;
    var ticker = function() {
        this.tick_timeout = setTimeout(ticker, self.tick_time);
        tetris_controller.cmd_down();
    };
    stdin.on('data', function(key) {
        tetris_controller.process_key(key);
    });
    ticker();
}

new TetrisInputProcessor().run();

File: Tetris.java
import java.io.Console;
import java.io.IOException;
import java.io.Reader;
import java.util.Random;
import java.util.List;
import java.util.ArrayList;

public class Tetris {
    public static final int PLAYFIELD_W = 10;
    public static final int PLAYFIELD_H = 20;
    public static final int PLAYFIELD_X = 30;
    public static final int PLAYFIELD_Y = 1;
    public static final TetrisColor BORDER_COLOR = TetrisColor.YELLOW;

    public static final int HELP_X = 58;
    public static final int HELP_Y = 1;
    public static final TetrisColor HELP_COLOR = TetrisColor.CYAN;

    public static final int SCORE_X = 1;
    public static final int SCORE_Y = 2;
    public static final TetrisColor SCORE_COLOR = TetrisColor.GREEN;

    public static final int NEXT_X = 14;
    public static final int NEXT_Y = 11;

    public static final int GAMEOVER_X = 1;
    public static final int GAMEOVER_Y = PLAYFIELD_H + 3;

    public static final int INITIAL_MOVE_DOWN_DELAY = 1000;
    public static final double DELAY_FACTOR = 0.8;
    public static final int LEVEL_UP = 20;

    public static final String NEXT_EMPTY_CELL = "  ";
    public static final String PLAYFIELD_EMPTY_CELL = " .";
    public static final String FILLED_CELL = "[]";

    public static final Random RANDOM = new Random();

    public static void main(String[] args) {
        try {
            new Tetris().run();
        } catch (IOException ioe) {
        } catch (InterruptedException ie) {
        }
    }

    public void run() throws IOException, InterruptedException {
        try {
            String[] cmd = {"/bin/sh", "-c", "stty raw -echo </dev/tty"};
            Runtime.getRuntime().exec(cmd).waitFor();
            Console console = System.console();
            Reader reader = console.reader();
            TetrisScreen screen = new TetrisScreen();
            TetrisController tc = new TetrisController(screen);
            Object lock = new Object();
            TetrisTicker tt = new TetrisTicker(tc, lock);
            tt.start();
            char key[] = {0, 0, 0};
            while (tc.running) {
                key[2] = key[1];
                key[1] = key[0];
                if (key[2] == 27 && key[1] == '[') {
                    key[0] = (char)reader.read();
                } else {
                    key[0] = Character.toLowerCase((char)reader.read());
                }
                synchronized(lock) {
                    switch(key[0]) {
                        case 3:
                        case 'q':
                            tc.cmdQuit();
                            break;
                        case 'C':
                        case 'd':
                            tc.cmdRight();
                            break;
                        case 'D':
                        case 'a':
                            tc.cmdLeft();
                            break;
                        case 'A':
                        case 's':
                            tc.cmdRotate();
                            break;
                        case ' ':
                            tc.cmdDrop();
                            break;
                        case 'h':
                            tc.cmdToggleHelp();
                            break;
                        case 'n':
                            tc.cmdToggleNext();
                            break;
                        case 'c':
                            tc.cmdToggleColor();
                            break;
                        default:
                            break;
                    }
                    screen.flush();
                }
            }
            tt.shutdown();
        } finally {
            String[] cmd = new String[] {"/bin/sh", "-c", "stty sane </dev/tty"};
            Runtime.getRuntime().exec(cmd).waitFor();
        }
    }
}

class TetrisTicker extends Thread {
    private static int delay = Tetris.INITIAL_MOVE_DOWN_DELAY;
    private TetrisController tc = null;
    private Object lock = null;
    private boolean go = true;

    public TetrisTicker(TetrisController tc, Object lock) {
        this.tc = tc;
        this.lock = lock;
    }

    public static void decreaseDelay() {
        delay *= Tetris.DELAY_FACTOR;
    }

    public void shutdown() {
        go = false;
    }

    public void run() {
        while (go) {
            try {
                Thread.sleep(delay);
            } catch (java.lang.InterruptedException ie) {
            }
            if (go) {
                synchronized(lock) {
                    tc.cmdDown();
                    tc.screenFlush();
                }
            }
        }
    }
}

enum TetrisColor {
    RED (1),
    GREEN (2),
    YELLOW (3),
    BLUE (4),
    FUCHSIA (5),
    CYAN (6),
    WHITE (7);

    public final int value;

    TetrisColor(int value) {
        this.value = value;
    }

    private static final TetrisColor VALUES[] = values();

    public static TetrisColor getRandomColor() {
        return VALUES[Tetris.RANDOM.nextInt(VALUES.length)];
    }
}

class TetrisScreen {
    private boolean useColor = true;
    private String s = "";

    public void print(String s) {
        this.s += s;
    }

    public void xyprint(int x, int y, String s) {
        this.s += ("\u001B[" + y + ";" + x + "H" + s);
    }

    public void showCursor() {
        s += "\u001B[?25h";
    }

    public void hideCursor() {
        s += "\u001B[?25l";
    }

    public void setFg(TetrisColor c) {
        if (useColor) {
            s += ("\u001B[3" + c.value + "m");
        }
    }

    public void setBg(TetrisColor c) {
        if (useColor) {
            s += ("\u001B[4" + c.value + "m");
        }
    }

    public void resetColors() {
        s += "\u001B[0m";
    }

    public void setBold() {
        s += "\u001B[1m";
    }

    public void clearScreen() {
        s += "\u001B[2J";
    }

    public void flush() {
        System.out.print(s);
        s = "";
    }

    public void toggleColor() {
        useColor ^= true;
    }
}

abstract class TetrisScreenItem {
    public boolean visible = true;
    protected TetrisScreen screen = null;

    public abstract void draw(boolean visible);

    public TetrisScreenItem(TetrisScreen screen) {
        visible = true;
        this.screen = screen;
    }

    public void show() {
        if (visible) {
            draw(true);
        }
    }

    public void hide() {
        if (visible) {
            draw(false);
        }
    }

    public void toggle() {
        visible ^= true;
        draw(visible);
    }
}

class TetrisHelp extends TetrisScreenItem {
    private final TetrisColor color = Tetris.HELP_COLOR;
    private String[] text = {
        "  Use cursor keys",
        "       or",
        "    s: rotate",
        "a: left,  d: right",
        "    space: drop",
        "      q: quit",
        "  c: toggle color",
        "n: toggle show next",
        "h: toggle this help"
    };

    public TetrisHelp(TetrisScreen screen) {
        super(screen);
    }

    public void draw(boolean visible) {
        screen.setBold();
        screen.setFg(color);
        for (int i = 0; i < text.length; i++) {
            String s = text[i];
            if (! visible) {
                s = new String(new char[s.length()]).replace("\0", " ");
            }
            screen.xyprint(Tetris.HELP_X, Tetris.HELP_Y + i, s);
        }
        screen.resetColors();
    }
}

class TetrisScore {
    private int score = 0;
    private int level = 1;
    private int linesCompleted = 0;
    private TetrisScreen screen = null;

    TetrisScore(TetrisScreen screen) {
        this.screen = screen;
    }

    public void update(int completeLines) {
        linesCompleted += completeLines;
        score += (completeLines * completeLines);
        if (score > Tetris.LEVEL_UP * level) {
            level += 1;
            TetrisTicker.decreaseDelay();
        }
        show();
    }

    public void show() {
        screen.setBold();
        screen.setFg(Tetris.SCORE_COLOR);
        screen.xyprint(Tetris.SCORE_X, Tetris.SCORE_Y,     "Lines completed: " + linesCompleted);
        screen.xyprint(Tetris.SCORE_X, Tetris.SCORE_Y + 1, "Level:           " + level);
        screen.xyprint(Tetris.SCORE_X, Tetris.SCORE_Y + 2, "Score:           " + score);
        screen.resetColors();
    }
}

class TetrisPlayField {
    private TetrisScreen screen = null;
    private List<List<TetrisColor>> cells = new ArrayList<List<TetrisColor>>();

    public TetrisPlayField(TetrisScreen screen) {
        this.screen = screen;
        for (int i = 0; i < Tetris.PLAYFIELD_H; i++) {
            cells.add(getEmptyRow());
        }
    }

    public List<TetrisColor> getEmptyRow() {
        List<TetrisColor> row = new ArrayList<TetrisColor>();
        for (int i = 0; i < Tetris.PLAYFIELD_W; i++) {
            row.add(null);
        }
        return row;
    }

    public void show() {
        for (int y = 0; y < cells.size(); y++ ) {
            List<TetrisColor> row = cells.get(y);
            screen.xyprint(Tetris.PLAYFIELD_X, Tetris.PLAYFIELD_Y + y, "");
            for (TetrisColor cell : row) {
                if (cell == null) {
                    screen.print(Tetris.PLAYFIELD_EMPTY_CELL);
                } else {
                    screen.setFg(cell);
                    screen.setBg(cell);
                    screen.print(Tetris.FILLED_CELL);
                    screen.resetColors();
                }
            }
        }
    }

    public void flattenPiece(TetrisPiece piece) {
        for (int[] cell : piece.getCells(null)) {
            int x = cell[0];
            int y = cell[1];
            cells.get(y).set(x, piece.color);
        }
    }

    public int processCompleteLines() {
        List<List<TetrisColor>> newCells = new ArrayList<List<TetrisColor>>();
        for (List<TetrisColor> row : cells) {
            if (row.indexOf(null) != -1) {
                newCells.add(row);
            }
        }
        int completeLines = Tetris.PLAYFIELD_H - newCells.size();
        for (int i = 0; i < completeLines; i++) {
            newCells.add(0, getEmptyRow());
        }
        cells = newCells;
        return completeLines;
    }

    public void drawBorder() {
        screen.setBold();
        screen.setFg(Tetris.BORDER_COLOR);
        for (int y = 0; y < Tetris.PLAYFIELD_H; y++) {
            // 2 because border is 2 characters thick
            screen.xyprint(Tetris.PLAYFIELD_X - 2, y + Tetris.PLAYFIELD_Y, "<|");
            // 2 because each cell on play field is 2 characters wide
            screen.xyprint(Tetris.PLAYFIELD_X + Tetris.PLAYFIELD_W * 2, y + Tetris.PLAYFIELD_Y, "|>");
        }

        screen.xyprint(Tetris.PLAYFIELD_X, Tetris.PLAYFIELD_Y + Tetris.PLAYFIELD_H, new String(new char[Tetris.PLAYFIELD_W]).replace("\0", "=="));
        screen.xyprint(Tetris.PLAYFIELD_X, Tetris.PLAYFIELD_Y + Tetris.PLAYFIELD_H + 1, new String(new char[Tetris.PLAYFIELD_W]).replace("\0", "\\/"));
        screen.resetColors();
    }

    public boolean positionOk(TetrisPiece piece, int[] position) {
        for (int[] cell : piece.getCells(position)) {
            int x = cell[0];
            int y = cell[1];
            if (x < 0 || x >= Tetris.PLAYFIELD_W || y < 0 || y >= Tetris.PLAYFIELD_H || cells.get(y).get(x) != null) {
                return false;
            }
        }
        return true;
    }
}

class TetrisPiece extends TetrisScreenItem {
    public TetrisColor color = null;
    public String emptyCell = Tetris.NEXT_EMPTY_CELL;
    public int[] origin = {0, 0};

    private int symmetry = 0;
    private int[] position = {0, 0, 0};
    private int[] data = null;
    // 0123
    // 4567
    // 89ab
    // cdef
    private static int[][] pieceData = {
        {0x1256}, // square
        {0x159d, 0x4567}, // line
        {0x4512, 0x0459}, // s
        {0x0156, 0x1548}, // z
        {0x159a, 0x8456, 0x0159, 0x2654}, // l
        {0x1598, 0x0456, 0x2159, 0xa654}, // inverted l
        {0x1456, 0x1596, 0x4569, 0x4159}  // t
    };

    public TetrisPiece(TetrisScreen screen, int[] origin, boolean visible) {
        super(screen);
        this.origin = origin;
        this.visible = visible;
        color = TetrisColor.getRandomColor();
        data = pieceData[Tetris.RANDOM.nextInt(pieceData.length)];
        symmetry = data.length;
        position = new int[]{0, 0, Tetris.RANDOM.nextInt(symmetry)};
    }

    public int[][] getCells(int[] newPosition) {
        int x = position[0];
        int y = position[1];
        int z = position[2];
        if (newPosition != null) {
            x = newPosition[0];
            y = newPosition[1];
            z = newPosition[2];
        }
        int currentData = data[z];
        int result[][] = {{0, 0}, {0, 0}, {0, 0}, {0, 0}};
        for (int i = 0; i < 4; i++) {
            result[i][0] = x + ((currentData >> 4 * i) & 3);
            result[i][1] = y + ((currentData >> 4 * i + 2) & 3);
        }
        return result;
    }

    public void draw(boolean visible) {
        if (visible) {
            screen.setFg(color);
            screen.setBg(color);
        }
        int ox = origin[0];
        int oy = origin[1];
        for (int[] cell : getCells(null)) {
            int x = cell[0];
            int y = cell[1];
            screen.xyprint(ox + x * 2, oy + y, visible ? Tetris.FILLED_CELL : emptyCell);
        }
        screen.resetColors();
    }

    public void setPosition(int[] p) {
        position = new int[]{p[0], p[1], p[2] < 0 ? position[2] : p[2]};
    }

    public int[] newPosition(int dx, int dy, int dz) {
        int x = position[0];
        int y = position[1];
        int z = position[2];
        return new int[]{x + dx, y + dy, (z + dz) % symmetry};
    }
}

class TetrisController {
    public boolean running = true;

    private TetrisScreen screen = null;
    private TetrisHelp help = null;
    private TetrisScore score = null;
    private TetrisPlayField playfield = null;
    private boolean nextPieceVisible = true;
    private TetrisPiece nextPiece = null;
    private TetrisPiece currentPiece = null;

    public TetrisController(TetrisScreen screen) {
        this.screen = screen;
        help = new TetrisHelp(screen);
        score = new TetrisScore(screen);
        playfield = new TetrisPlayField(screen);
        getNextPiece();
        getCurrentPiece();
        redrawScreen();
        screen.flush();
    }

    public void getCurrentPiece() {
        nextPiece.hide();
        currentPiece = nextPiece;
        currentPiece.setPosition(new int[]{(Tetris.PLAYFIELD_W - 4) / 2, 0, -1});
        if (! playfield.positionOk(currentPiece, null)) {
            cmdQuit();
            return;
        }
        currentPiece.visible = true;
        currentPiece.emptyCell = Tetris.PLAYFIELD_EMPTY_CELL;
        currentPiece.origin = new int[]{Tetris.PLAYFIELD_X, Tetris.PLAYFIELD_Y};
        currentPiece.show();
        getNextPiece();
    }

    public void getNextPiece() {
        nextPiece = new TetrisPiece(screen, new int[]{Tetris.NEXT_X, Tetris.NEXT_Y}, nextPieceVisible);
        nextPiece.show();
    }

    public void cmdToggleColor() {
        screen.toggleColor();
        redrawScreen();
    }

    public void cmdToggleNext() {
        nextPieceVisible ^= true;
        nextPiece.toggle();
    }

    public void cmdToggleHelp() {
        help.toggle();
    }

    public void cmdQuit() {
        running = false;
        screen.xyprint(Tetris.GAMEOVER_X, Tetris.GAMEOVER_Y, "Game over!");
        screen.xyprint(Tetris.GAMEOVER_X, Tetris.GAMEOVER_Y + 1, "");
        screen.showCursor();
    }

    public void cmdRotate() {
        move(0, 0, 1);
    }

    public void cmdLeft() {
        move(-1, 0, 0);
    }

    public void cmdRight() {
        move(1, 0, 0);
    }

    public boolean cmdDown() {
        if (move(0, 1, 0)) {
            return true;
        }
        getCurrentPiece();
        return false;
    }

    public void cmdDrop() {
        while (cmdDown()) {};
    }

    public void redrawScreen() {
        screen.clearScreen();
        screen.hideCursor();
        playfield.drawBorder();
        help.show();
        playfield.show();
        score.show();
        nextPiece.show();
        currentPiece.show();
    }

    public void processFallenPiece() {
        playfield.flattenPiece(currentPiece);
        int completeLines = playfield.processCompleteLines();
        if (completeLines > 0) {
            score.update(completeLines);
            playfield.show();
        }
    }

    public boolean move(int dx, int dy, int dz) {
        int[] newPosition = currentPiece.newPosition(dx, dy, dz);
        if (playfield.positionOk(currentPiece, newPosition)) {
            currentPiece.hide();
            currentPiece.setPosition(newPosition);
            currentPiece.show();
            return true;
        }
        if (dy == 0) {
            return true;
        }
        processFallenPiece();
        return false;
    }

    public void screenFlush() {
        screen.flush();
    }
}

